ESQL: Disallow remote enrich after lu join

alex-spies opened the pull request:
Fix https://github.com/elastic/elasticsearch/issues/129372

Due to how remote ENRICH is [planned](https://github.com/elastic/elasticsearch/blob/32e50d0d94e27ee559d24bf9d5463ba6e64d1788/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/mapper/Mapper.java#L93), it interacts in special ways with pipeline breakers, in particular LIMIT and TopN; when these are encountered upstream from a remote ENRICH, these nodes are copied and executed a second time after the remote ENRICH.

We'd like to allow remote ENRICH after LOOKUP JOIN, but that forces the lookup to be remote as well; this has its own interactions with pipeline breakers: in particular, LIMITs and TopNs cannot just be duplicated after LOOKUP JOIN, as LOOKUP JOIN may add new rows.

For now, let's just forbid any usage of remote ENRICH after LOOKUP JOINs; remote ENRICH is mostly relevant for CCS, and LOOKUP JOIN doesn't support that in 9.1/8.19, anyway.

There is separate work that enables remote LOOKUP JOINs on remote clusters and adds the correct validations; we can later build support for remote ENRICH + LOOKUP JOIN on top of that. (C.f. my comment [here](https://github.com/elastic/elasticsearch/issues/129372#issuecomment-3083024230) and my draft https://github.com/elastic/elasticsearch/pull/131286 for enabling this.)

alex-spies committed:
Disallow LOOKUP JOIN before remote ENRICH

alex-spies committed:
Add some csv tests

alex-spies committed:
Update docs/changelog/131426.yaml

alex-spies committed:
Make verification work with multiple FORK branches

This was broken before: a coordinator-only ENRICH in one branch and a
remote-only ENRICH in the other could be (wrongly) considered invalid
depending on the order of the fork branches.

alex-spies committed:
Exclude new tests from running in CCS

This interprets LOOKUP JOIN as a remote-only LOOKUP JOIN which is more
restricted.

alex-spies committed:
Add comment

luigidellaquila reviewed (approved):
Thanks Alex, LGTM.

I left a few comments, especially about improvements we can do in test coverage.

One more thing we can do is add `ENRICH _remote:` to the generative tests, but that's something we can do in a follow-up PR

smalyshev commented:
LGTM in general, added some thoughts & nitpicks inline.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -2343,7 +2348,60 @@ public void testRemoteLookupJoinIsDisabled() {
+        String err = error("""
+            FROM test
+            | EVAL language_code = languages
+            | LOOKUP JOIN languages_lookup ON language_code
+            | ENRICH _remote:languages ON language_code
```
One approach of making the remote ENRICH planning sane is to implement an optimizer rule that tries to push down remote ENRICH super aggressively, so it ends up on the data node naturally. Not sure it's the one we'll settle on, though, but we'll probably not backport it to 9.1.

But more test cases is always better :) Nice observation! I'll add some.

alex-spies committed:
Add tests: remote ENRICH + STATS/coord ENRICH

alex-spies committed:
More ENRICH csv tests for remote/coord mode

alex-spies committed:
Also use non-overwriting LU JOIN in VerifierTests

Randomly use a lookup index that doesn't shadow the columns added by the
ENRICH. This could be relevant if we at some point allow LU JOIN after
remote ENRICH in the edge case that the two can be swapped around.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into disallow-remote-enrich-after-lu-join

alex-spies committed:
Make lambda more readable

alex-spies commented:
Thanks for the reviews @luigidellaquila and @smalyshev , very helpful feedback!

alex-spies committed:
Fix the added randomized verifier tests

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into disallow-remote-enrich-after-lu-join

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.19  | Commit could not be cherrypicked due to conflicts |
| ❌ |  9.1  | Commit could not be cherrypicked due to conflicts |
| ❌ |  9.0  | Commit could not be cherrypicked due to conflicts |
| ❌ |  8.18  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 131426`

alex-spies commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/131533">](https://github.com/elastic/elasticsearch/pull/131533)|
|✅|9.0|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/131534">](https://github.com/elastic/elasticsearch/pull/131534)|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/131535">](https://github.com/elastic/elasticsearch/pull/131535)|
|✅|8.18|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/131537">](https://github.com/elastic/elasticsearch/pull/131537)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
